### ISSUE TITLE
Appservice Sender: Maximum recursion depth exceeded

### ISSUE THREAD

**Issue by turt2live** [matrix-org/synapse#3539]
Unknown cause. Happened on t2bot.io while the Telegram bridge caught up on hundreds of transactions (at a rate of ~8Hz for 20 minutes).

```
appservice - 2018-07-16 04:16:51,096 - twisted - 131 - CRITICAL - - Unhandled error in Deferred:
appservice - 2018-07-16 04:16:51,097 - twisted - 131 - CRITICAL - -
Traceback (most recent call last):
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 654, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1475, in gotResult
    _inlineCallbacks(r, g, status)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1421, in _inlineCallbacks
    status.deferred.callback(getattr(e, "value", None))
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 460, in callback
    self._startRunCallbacks(result)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 568, in _startRunCallbacks
    self._runCallbacks()
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 667, in _runCallbacks
    current.result = failure.Failure(captureVars=self.debug)
RuntimeError: maximum recursion depth exceeded
```

**Comment by turt2live** [comment]
I'm seeing this a lot more often now :(

```
appservice_1 - 2018-11-28 23:39:19,070 - twisted - 244 - CRITICAL - - Unhandled error in Deferred:
appservice_1 - 2018-11-28 23:39:19,074 - twisted - 244 - CRITICAL - -
Traceback (most recent call last):
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 654, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1475, in gotResult
    _inlineCallbacks(r, g, status)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1421, in _inlineCallbacks
    status.deferred.callback(getattr(e, "value", None))
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 460, in callback
    self._startRunCallbacks(result)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 568, in _startRunCallbacks
    self._runCallbacks()
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 667, in _runCallbacks
    current.result = failure.Failure(captureVars=self.debug)
RuntimeError: maximum recursion depth exceeded
```

**Comment by richvdh** [comment]
dup of #4336 I think, which was fixed by #5885
